x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/scroll/ScrollDataExtractor.java

```diff
@@ -213,23 +222,28 @@ private InputStream processAndConsumeSearchHits(SearchHit hits[]) throws IOExcep
 
     private InputStream continueScroll() throws IOException {
         logger.debug("[{}] Continuing scroll with id [{}]", context.jobId, scrollId);
-        SearchResponse searchResponse;
+        SearchResponse searchResponse = null;
         try {
-            searchResponse = executeSearchScrollRequest(scrollId);
-        } catch (SearchPhaseExecutionException searchExecutionException) {
-            if (searchHasShardFailure) {
-                throw searchExecutionException;
+            try {
+                searchResponse = executeSearchScrollRequest(scrollId);
+            } catch (SearchPhaseExecutionException searchExecutionException) {
+                if (searchHasShardFailure) {
+                    throw searchExecutionException;
+                }
+                logger.debug("[{}] search failed due to SearchPhaseExecutionException. Will attempt again with new scroll", context.jobId);
+                markScrollAsErrored();
+                searchResponse = executeSearchRequest(buildSearchRequest(lastTimestamp == null ? context.start : lastTimestamp));
+            }
+            logger.debug("[{}] Search response was obtained", context.jobId);
+            timingStatsReporter.reportSearchDuration(searchResponse.getTook());
+            scrollId = searchResponse.getScrollId();
+            SearchHit hits[] = searchResponse.getHits().getHits();
+            return processAndConsumeSearchHits(hits);
```

Comment:
I think `processAndConsumeSearchHits(hits);` actually consumes the hits and thus a reference isn't required any longer (its not async).

@droberts195 could you confirm?

Reply:
@original-brownbear OK, thank you for clarification. I misunderstood as I thought the whole point of ref counting was allowing hits to re-use buffered bytes.

Reply:
>  of ref counting was allowing hits to re-use buffered bytes.

Right but we want granularity at the `SearchHit` level here in the end, otherwise holding on to a single hit could hold on to a much much larger buffer containing other hits in the same response. 
We need the ref-counting on the search response to enable ref-counting search hits at the level of these changes but each `SearchHit` itself will be tracked individually in the end.

Reply:
I guess the question was about whether this code is dangerous now: https://github.com/elastic/elasticsearch/blob/d68670fb88586ca7dc14fb57da4380faf226bbcb/x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/scroll/ScrollDataExtractor.java#L206-L208

The reason that's there is that as we iterate the hits we're building up a similar sized list of different objects derived from the hits. If garbage collection needs to run half way through iterating the hits we want it to be able to reclaim the half that we've already processed. Otherwise as we near the end of the hits we'll be temporarily using around double the memory of the hits.

Reply:
It will become dangerous once the search hits are ref counted, but we won't miss that with the leak tracking infrastructure in place. I'll address this in the PR that makes the hits ref counted, nothing to worry about yet :)